Feature/expiration headers

dariuszor opened the pull request:
@butenkor could you have a look please?

dariuszor commented:
@hajoeichler Hi, could you take a look?
It's pretty urgent for merkur. Also: should I bump a version?

dariuszor commented:
@onibox, @anas-aso, maybe you could haave a look?

hajoeichler commented on src/coffee/services/s3client.coffee:
```diff
@@ -220,10 +220,13 @@ class S3Client
         @sendMetrics 'increment', 'image.resized'
-        header = 'x-amz-acl': 'public-read'
+        headers = 'x-amz-acl': 'public-read'
+        if !isNaN(expire)
```
use `_.isNumber`

hajoeichler commented on src/coffee/services/s3client.coffee:
```diff
@@ -220,10 +220,13 @@ class S3Client
+        headers = 'x-amz-acl': 'public-read'
+        if !isNaN(expire)
+          headers['Cache-Control'] = 'max-age=' + expire + ', public'
+          headers['Expires'] = 'Sun, 01 Jan 2034 00:00:00 GMT'
```
I don't like this hard coded value.

dariuszor replied:
There was one source which suggested to set both Cache-Control and Expires.
C-C takes precedence over Expires.
So this Expires header was set ‘just in case’ and it is safe date.
It was pretty old comment, when C-C wasn't that long on the market. 
Right now using C-C should be enough for this purposes. 
I will remove Expires header.

hajoeichler commented on src/spec/services/s3client.spec.coffee:
```diff
@@ -124,25 +124,43 @@ describe 'S3Client', ->
       .then -> done()
       .catch (err) -> done(err)
+
+    it 'should call put function without \"extends\" paramterer', (done) ->
```
make the test description more clear.
"should call put without cache control headers"

hajoeichler commented:
A general question. Doesn't it make sense to always set a cache header, even if not provided the tool defines the default value.

dariuszor commented:
@hajoeichler yes, it makes sense. Which value should I use as a default? 30 days?

emmenko commented on src/coffee/services/s3client.coffee:
```diff
@@ -220,10 +220,14 @@ class S3Client
         @sendMetrics 'increment', 'image.resized'
-        header = 'x-amz-acl': 'public-read'
+        if isNaN expire
+          expire = 2592000
```
What do you need this for?

dariuszor replied:
For merkur we need to set cache-control, max-age header for images for optimization.
After setting the expiration headers a browser will only make requests to the server if it does not have the resource in its cache.

emmenko replied:
No I mean to set the number here. If it's a default, you should set it in the command option itself. And also what does this number mean? I guess it's 30days, but add a comment next to it with the days value.

PS: please don't mention customer names, it's a public repo!

dariuszor replied:
Oh, sorry. Edited. I will update it acordingly. Thanks for feedback!

svenmueller commented on README.md:
```diff
@@ -192,6 +192,7 @@ Options:
 - `descriptions <path>` - Required
 - `regex [name]` - Optional
 - `logFile <path>` - Optional
+- `expire <value>` - Optional
```
compress/expire should not be added as an option here. instead it should be possible to configure it in the description yaml (this file describes the handling of the AWS S3 resources!). Also please add proper info in the README, since `expire` and `compress` doesn't give any info.

svenmueller commented on src/coffee/commands/images-convert.coffee:
```diff
@@ -12,6 +12,7 @@ program
 .option '-d, --descriptions <path>', 'set image descriptions file path'
 .option '-r, --regex [name]', 'an optional RegExp used for filtering listed files (e.g.: /(.*)\.jpg/)', ''
 .option '-l, --logFile <path>', 'optionally log to a file instead of printing to console (errors will still be printed to stderr)'
+.option '-e, --expire <value>', 'set an expiration header - Cache-control: max-age. Default 30 days.', parseInt
```
Please move configuration for compress/expire headers to the AWS S3 description yaml file

svenmueller commented:
@dariuszor please see my inline comments

svenmueller commented:
Btw, we are already using this tooling in many projects, so the behavior should not change for them (eg. no expiration header, images are not compressed etc) if not configured.

dariuszor commented:
@svenmueller thanks for detailed infos! 
I will move configuration to yaml file, update comments/README and remove default values if were not configured.

svenmueller commented:
Thx a lot :) :+1:

dariuszor commented:
@svenmueller I've updated PR. I hope I didn't miss anything.
I removed one test from previous commit. I didn't feel it test anything, and it hadn't influence on coverage either.
`headers = 'x-amz-acl': 'public-read'` was implemented before, so I left it to not change configuration of other project.

svenmueller commented:
@dariuszor LGTM

dariuszor commented:
@svenmueller Great, thanks!
Coudl you merge? I don't have write access.
